### PR TITLE
I created the Service

### PR DESCRIPTION
**Pull request by chaunguyen184**
In the Service class, there are a lot of errors about "protect" accessing. Can you take a look about that

### CODE REVIEW

File: src/main/java/sm/rental/model/SMRental.java [hunk 1]
```diff
@@ -31,13 +32,14 @@
 	// Define the reference variables to the various 
 	// entities with scope Set and Unary
 
-	@Getter private ArrayList<Van> rgVans;
-	@Getter private RentalCounter rgRentalCounter;
+	@Getter protected ArrayList<Van> rgVans;
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/SMRental.java]: Why did you change this? Please revert lines 34 to 41 to the originals. These are the reasons why your code didn't build properly.

File: src/main/java/sm/rental/model/SMRental.java [hunk 2]
```diff
@@ -31,13 +32,14 @@
...
+	@Getter protected RentalCounter rgRentalCounter;
+
+	@Getter protected ArrayList<Customer> Terminal1 = new ArrayList<Customer>();
+	@Getter protected ArrayList<Customer> Terminal2 = new ArrayList<Customer>();
+	@Getter protected ArrayList<Customer> qReturnLine = new ArrayList<Customer>();
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/SMRental.java]: The name change I can understand but can you be consistent and apply it to the terminals as well? 

While you are changing this, can you put the terminals in a set?

File: src/main/java/sm/rental/model/activities/Service.java [hunk 3]
```diff
@@ -1,4 +1,50 @@
...
+    @Getter private Customer icCustomer;
+    @Getter private SMRental model;
+
+    public Service(SMRental model){
+        this.model = model;
+    }
+
+    protected static boolean precondition(SMRental model){
+        boolean result = false;
+        int numOfCustomerAtCounter = model.rgRentalCounter.getN();
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: please use the getters. 

If you are getting errors on `model.getRentalLine()`, then let me know and I'll help you configure lombok properly. thanks.

File: src/main/java/sm/rental/model/activities/Service.java [hunk 4]
```diff
@@ -1,4 +1,50 @@
...
+        }
+        return result;
+    }
+
+    protected double duration(){
+        return (this.model.rvp.uServiceTime(this.icCustomer.uType));
+    }
+
+    public void startingEvent(){
+        model.rgRentalCounter.getN()++;
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: use the getters

File: src/main/java/sm/rental/model/activities/Service.java [hunk 5]
```diff
@@ -1,4 +1,50 @@
...
+        icCustomer = model.qRentalLine.remove(0); // Remove customer
+    }
+
+    public void terminatingEvent(){
+        Output output = this.model.output;
+        model.rgRentalCounter.getN()--;
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: there is no rgRentalCounter.getN() function. I added free agent, and occupy agent. Please use these

File: src/main/java/sm/rental/model/activities/Service.java [hunk 6]
```diff
@@ -1,4 +1,50 @@
...
+        if(this.icCustomer.uType == Customer.CustomerType.NEW){
+            if((model.getClock() - icCustomer.startTime)<= TARGET_TURNARROUND_NEW){
+                output.numSatisfied++;
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: output needs to be defined before we can arbitrary outputs. For now leave it but we'll need to revisit this code.

File: src/main/java/sm/rental/model/activities/Service.java [hunk 7]
```diff
@@ -1,4 +1,50 @@
...
+        return result;
+    }
+
+    protected double duration(){
+        return (this.model.rvp.uServiceTime(this.icCustomer.uType));
+    }
+
+    public void startingEvent(){
+        model.rgRentalCounter.getN()++;
+        icCustomer = model.qRentalLine.remove(0); // Remove customer
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: use the getters.

File: src/main/java/sm/rental/model/activities/Service.java [hunk 8]
```diff
@@ -1,4 +1,50 @@
...
+    }
+
+    public void terminatingEvent(){
+        Output output = this.model.output;
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: We should have a getter in SMRental, add it and use it.

File: src/main/java/sm/rental/model/activities/Service.java [hunk 9]
```diff
@@ -1,4 +1,50 @@
...
+        model.rgRentalCounter.getN()--;
+        if(this.icCustomer.uType == Customer.CustomerType.NEW){
+            if((model.getClock() - icCustomer.startTime)<= TARGET_TURNARROUND_NEW){
+                output.numSatisfied++;
+            }
+            output.numServed++;
+            output.customerSatisfactionRate = output.numSatisfied / output.numServed;
+        }
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: can use an else

File: src/main/java/sm/rental/model/SMRental.java [hunk 10]
```diff
@@ -32,26 +32,26 @@
...
+	@Getter private ArrayList<Van> rgVans;
+	@Getter private RentalCounter rgRentalCounter;
 
-	@Getter protected ArrayList<Customer> Terminal1 = new ArrayList<Customer>();
-	@Getter protected ArrayList<Customer> Terminal2 = new ArrayList<Customer>();
-	@Getter protected ArrayList<Customer> qReturnLine = new ArrayList<Customer>();
-	@Getter protected ArrayList<Customer> qRentalLine = new ArrayList<Customer>();
+	@Getter private ArrayList<Customer> Terminal1 = new ArrayList<Customer>();
+	@Getter private ArrayList<Customer> Terminal2 = new ArrayList<Customer>();
+	@Getter private ArrayList<Customer> qReturnLine = new ArrayList<Customer>();
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/SMRental.java]: Rename these to the original name `Returnline` then right click, hit refactor, hit rename then rename it. That will fix every reference to it...

File: src/main/java/sm/rental/model/activities/Service.java [hunk 11]
```diff
@@ -14,26 +15,26 @@ public Service(SMRental model){
 
     protected static boolean precondition(SMRental model){
         boolean result = false;
-        int numOfCustomerAtCounter = model.rgRentalCounter.getN();
-        int numOfAgents = model.rgRentalCounter.uNumAgents;
-        if((model.qRentalLine.size() != 0) && (numOfCustomerAtCounter < numOfAgents)){
+        int numOfCustomerAtCounter = model.getRgRentalCounter().getN();
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: Again, there is not `RentalCounter.getN();`

File: src/main/java/sm/rental/model/activities/Service.java [hunk 12]
```diff
@@ -14,26 +15,26 @@ public Service(SMRental model){
...
 
     protected double duration(){
-        return (this.model.rvp.uServiceTime(this.icCustomer.uType));
+        return (this.model.getRvp().uServiceTime(this.icCustomer.uType));
     }
 
     public void startingEvent(){
-        model.rgRentalCounter.getN()++;
-        icCustomer = model.qRentalLine.remove(0); // Remove customer
+        model.getRgRentalCounter().getN();
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: Again, there is not `RentalCounter.getN();`

File: src/main/java/sm/rental/model/activities/Service.java [hunk 13]
```diff
@@ -14,26 +15,26 @@ public Service(SMRental model){
...
+        icCustomer = model.getQRentalLine().remove(0); // Remove customer
     }
 
     public void terminatingEvent(){
-        Output output = this.model.output;
-        model.rgRentalCounter.getN()--;
+        Output_1 output = this.model.getOutput1();
+        model.getRgRentalCounter().getN()--;
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: Again, I don't know what you think this is doing...

File: src/main/java/sm/rental/model/entities/Customer.java [hunk 14]
```diff
@@ -8,12 +8,14 @@
         RETURNING
     }
 
-    @Getter private double startTime; //The time at which a customer enters the system.
-    @Getter private CustomerType uType;
-    @Getter private int numPassengers; //The number of passengers accompanying the customer plus the customer themselves.
+    public double startTime; //The time at which a customer enters the system.
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/entities/Customer.java]: I don't know why you are changing this but this is a regression.

File: src/main/java/sm/rental/model/entities/RentalCounter.java [hunk 15]
```diff
@@ -1,8 +1,12 @@
 package  sm.rental.model.entities;
+import lombok.Getter;
+
+import java.util.ArrayList;
 
 public class RentalCounter {
 
-    private int uNumAgents; //The number of available agents.
+    @Getter private int uNumAgents; //The number of available agents.
+    @Getter private ArrayList<Customer> group = new ArrayList<Customer>();
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/entities/RentalCounter.java]: RentalCounter doesn't have an associated list, infact the professor told us to remove the list.

File: src/main/java/sm/rental/model/entities/RentalCounter.java [hunk 16]
```diff
@@ -14,7 +18,24 @@ public void occupyAgent() throws RuntimeException { // decrease the number of ag
         if(this.uNumAgents == 0) throw new RuntimeException();
         this.uNumAgents -= 1;
     }
+
     public void freeAgent(){ // increases the amount of available agents by one
         this.uNumAgents += 1;
     }
+
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/entities/RentalCounter.java]: All these changes:
- don't work the way you use them
- don't match the model
- aren't described in your commit message.

File: src/main/java/sm/rental/model/SMRental.java (hunk 1 above)
Comment (chaunguyen184) [reply src/main/java/sm/rental/model/SMRental.java]: I forgot to get it back to "private". But I didn't change anything else

File: src/main/java/sm/rental/model/SMRental.java (hunk 2 above)
Comment (chaunguyen184) [reply src/main/java/sm/rental/model/SMRental.java]: I didn't change anything else except the type of ArraylList. It should be "private", I forgot to change it. But it shouldn't make any different

File: src/main/java/sm/rental/model/activities/Service.java [hunk 17]
```diff
@@ -15,36 +17,27 @@ public Service(SMRental model){
 
     protected static boolean precondition(SMRental model){
         boolean result = false;
-        int numOfCustomerAtCounter = model.getRgRentalCounter().getN();
-        int numOfAgents = model.getRgRentalCounter().getUNumAgents();
-        if((model.getQRentalLine().size() != 0) && (numOfCustomerAtCounter < numOfAgents)){
+        if((model.getQRentalLine().size() > 0) && (model.getUdp().isAgentAvailable()==true)){
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: `model.getUdp().isAgentAvailable()==true)` should return a boolean `True` or `False` So you don't need to perform the check because `True && True` is `True`, and `True && False` is `False`

File: src/main/java/sm/rental/model/activities/Service.java [hunk 18]
```diff
@@ -15,36 +17,27 @@ public Service(SMRental model){
 
     protected static boolean precondition(SMRental model){
         boolean result = false;
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: This variable isn't useful. You can just as easily `return True;`

File: src/main/java/sm/rental/model/activities/Service.java [hunk 19]
```diff
@@ -15,36 +17,27 @@ public Service(SMRental model){
...
             result = true;
         }
         return result;
     }
-
-    protected double duration(){
-        return (this.model.getRvp().uServiceTime(this.icCustomer.uType));
+    public void startingEvent(){
+        this.model.getUdp().OccupyAgent();
+        this.model.getQRentalLine().remove(icgCustomer); // Remove customer
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: You need to remove the first customer and save a reference to it like so: 
`icgCustomer = model.getQRentalLine().pop()`
However, arraylist doesn't have a pop function so there are 2 options:
- We use a different data structure
- We implement our own pop

Lastly, we should verify that this would 100% of the time act as a FIFO.

File: src/main/java/sm/rental/model/activities/Service.java [hunk 20]
```diff
@@ -15,36 +17,27 @@ public Service(SMRental model){
...
     }
 
-    public void startingEvent(){
-        model.getRgRentalCounter().getN();
-        icCustomer = model.getQRentalLine().remove(0); // Remove customer
+    protected double duration(){
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: Why is this protected again?

File: src/main/java/sm/rental/model/activities/Service.java [hunk 21]
```diff
@@ -15,36 +17,27 @@ public Service(SMRental model){
...
-        Output_1 output = this.model.getOutput1();
-        model.getRgRentalCounter().getN()--;
-        if(this.icCustomer.uType == Customer.CustomerType.NEW){
-            if((model.getClock() - icCustomer.startTime)<= TARGET_TURNARROUND_NEW){
-                output.numSatisfied++;
-            }
-            output.numServed++;
-            output.customerSatisfactionRate = output.numSatisfied / output.numServed;
+        this.model.getUdp().FreeAgent();
+        if (this.icgCustomer.uType.equals("NEW")){
```
Comment (NuclearBanane) [inline src/main/java/sm/rental/model/activities/Service.java]: uType is an enumeration. The string, human-readable form is not saved anywhere so there is no way to go from "NEW" to Customer.CustomerType.NEW
Should be:
`icgCustomer.uType == Customer.CustomerType.new`
